[SPARK-4809] Rework Guava library shading.

vanzin opened the pull request:
The current way of shading Guava is a little problematic. Code that
depends on "spark-core" does not see the transitive dependency, yet
classes in "spark-core" actually depend on Guava. So it's a little
tricky to run unit tests that use spark-core classes, since you need
a compatible version of Guava in your dependencies when running the
tests. This can become a little tricky, and is kind of a bad user
experience.

This change modifies the way Guava is shaded so that it's applied
uniformly across the Spark build. This means Guava is shaded inside
spark-core itself, so that the dependency issues above are solved.
Aside from that, all Spark sub-modules have their Guava references
relocated, so that they refer to the relocated classes now packaged
inside spark-core. Before, this was only done by the time the assembly
was built, so projects that did not end up inside the assembly (such
as streaming backends) could still reference the original location
of Guava classes.

The Guava classes are added to the "first" artifact Spark generates
(network-common), so that all downstream modules have the needed
classes available. Since "network-common" is a dependency of spark-core,
all Spark apps should get the relocated classes automatically.

Commit:
[SPARK-4809] Rework Guava library shading.

The current way of shading Guava is a little problematic. Code that
depends on "spark-core" does not see the transitive dependency, yet
classes in "spark-core" actually depend on Guava. So it's a little
tricky to run unit tests that use spark-core classes, since you need
a compatible version of Guava in your dependencies when running the
tests. This can become a little tricky, and is kind of a bad user
experience.

This change modifies the way Guava is shaded so that it's applied
uniformly across the Spark build. This means Guava is shaded inside
spark-core itself, so that the dependency issues above are solved.
Aside from that, all Spark sub-modules have their Guava references
relocated, so that they refer to the relocated classes now packaged
inside spark-core. Before, this was only done by the time the assembly
was built, so project that did not end up inside the assembly (such
as streaming backends) could still reference the original location
of Guava classes).

This relocation does not apply to the sub-modules under network/,
though. For those cases, we want to keep the Guava dependency alive,
since we want to use the same Guava as the rest of the Yarn NM
when deploying the auxiliary shuffle service. For this reason, also,
the network/ dependencies are shaded into the spark-core artifact
too, so that the raw Guava dependency doesn't leak.

Commit:
Merge branch 'master' into SPARK-4809

Commit:
Add fix for SPARK-5052 (PR #3874).

Commit:
Merge branch 'master' into SPARK-4809

Conflicts:
	streaming/pom.xml

SparkQA commented:
[Test build #25133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25133/consoleFull) for   PR 3658 at commit [`8053dd4`](https://github.com/apache/spark/commit/8053dd4080283385b868c32554d6e103164e0db4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25133/consoleFull) for   PR 3658 at commit [`8053dd4`](https://github.com/apache/spark/commit/8053dd4080283385b868c32554d6e103164e0db4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25133/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-4809

Conflicts:
	examples/pom.xml

SparkQA commented:
[Test build #25442 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25442/consoleFull) for   PR 3658 at commit [`f78c48a`](https://github.com/apache/spark/commit/f78c48a9a52122839f657de19944a81783ff59e5).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25442 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25442/consoleFull) for   PR 3658 at commit [`f78c48a`](https://github.com/apache/spark/commit/f78c48a9a52122839f657de19944a81783ff59e5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25442/
Test FAILed.

Commit:
Merge branch 'master' into SPARK-4809

SparkQA commented:
[Test build #25477 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25477/consoleFull) for   PR 3658 at commit [`941848f`](https://github.com/apache/spark/commit/941848f409739331b1d495c4fadb57dc5c373ded).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25477 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25477/consoleFull) for   PR 3658 at commit [`941848f`](https://github.com/apache/spark/commit/941848f409739331b1d495c4fadb57dc5c373ded).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `    val classServer                                   = new HttpServer(conf, outputDir, new SecurityManager(conf), classServerPort, "HTTP class server")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25477/
Test PASSed.

vanzin commented:
@pwendell would appreciate some input here. Thanks!

vanzin commented:
Ping.

pwendell commented on core/pom.xml:
```diff
@@ -408,6 +372,20 @@
+            <includes>
+              <include>com.google.guava:guava</include>
+              <include>org.apache.spark:spark-network-common_${scala.binary.version}</include>
+              <include>org.apache.spark:spark-network-shuffle_${scala.binary.version}</include>
```
Is this going to inline other modules (spark-network-common and spark-network-shuffle) in the published spark-core jar?

vanzin replied:
Yes. From the first comment:

> For this reason, also, the network/ dependencies are shaded into the spark-core artifact too, so that the raw Guava dependency doesn't leak.

pwendell commented on pom.xml:
```diff
@@ -1276,6 +1279,19 @@
+              <pattern>com.google.common</pattern>
+              <shadedPattern>org.spark-project.guava</shadedPattern>
+              <excludes>
+                <exclude>com/google/common/base/Absent*</exclude>
```
mind adding a quick note here to explain that we expose these in our API so we don't relocate them?

pwendell commented:
Hey so a couple questions about this. This does make things simpler overall in terms of the build. Some higher level questions:

1. This adds 1,500 class files and about 2 MB or a 30% in file size to the Spark core jar. Does this suggest any limits on how many overall dependencies we can shade in Spark longer term (i.e. in addition to guava)?
2. Inlining two of our own sub-projects into the spark-core jar seems a bit strange. Why not just have the shaded guava in the little assembly jar we create for YARN? Is there a reason we _must_ use the guava dependency that is provided on the cluster? If we did take the current approach, does this mean that Spark core as advertised in maven central won't depend on Spark's network modules... would we still publish the network modules then?

vanzin commented:
> Does this suggest any limits on how many overall dependencies we can shade in Spark longer term (i.e. in addition to guava)?

If we choose relocation as the official way of fixing potential library version conflicts, we'll end up with a growing spark-core. I don't think there's a limit to how many classes you can add (SPARK-1520 notwithstanding), but it doesn't look very pretty, I agree. I did this mainly to fix some issues with the current way Guava is shaded, though, not as a blueprint for how to fix dependency issues going forward.

(It also can confuse certain IDEs that automatically add import statements...)

> Why not just have the shaded guava in the little assembly jar we create for YARN?

That's definitely an option. Should be pretty easy to do if that's the preferred way, but I remember that it was a conscious choice to depend on the Yarn-provided Guava.

> would we still publish the network modules then?

Do they provide any public APIs? If they do, then yes. Otherwise, there would be no need to publish them.

Commit:
Add comment.

SparkQA commented:
[Test build #26116 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26116/consoleFull) for   PR 3658 at commit [`b3104fc`](https://github.com/apache/spark/commit/b3104fcb12d5d85f7514f575beeb671816b02251).
 * This patch merges cleanly.

pwendell commented:
Yeah, so I think it would be better to just shade guava in YARN as well. The main dependency-related constraint was that we didn't want Spark's yarn shuffle service to have any external dependencies (i.e. it should work well out of the box with YARN). But having an inlined shaded dependency doesn't matter in this regard. So I don't see any other compelling reason to use YARN's guava.

Overall is easier to understand what is going on if we don't have a special case there. And inlining things that we expose as semi-public classes (there are some developer API's in those modules, IIRC), it's a bit off from the normal maven model. So for all these reasons I think it would be good to just shade it in that jar as well.

SparkQA commented:
[Test build #26116 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26116/consoleFull) for   PR 3658 at commit [`b3104fc`](https://github.com/apache/spark/commit/b3104fcb12d5d85f7514f575beeb671816b02251).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26116/
Test PASSed.

pwendell commented:
Hey @vanzin mind bringing this up to date now (I merged #3874).

vanzin commented:
Hey @pwendell, I'll try to get to this soon. But I wanted to get your feedback on my idea for fixing the `network/` dependencies thing before I try to implement it.

The way I see it, the cleanest way is to do the Guava shading in the earliest artifact possible; that would be `network/common`. So that artifact would have the honor of providing all the relocated Guava classes to everyone. Since `spark-core` depends on it, everything should work out.

The only downside I see to that is that `network/common` would now expose `Optional` and friends when it's not really its fault (`spark-core` demands it).

What do you think?

pwendell commented:
@vanzin yeah this is a fair point, this would mean that network/common would expose the (un-shaded) Guava classes... a bit clunky. If those classes change in the future we could get into trouble. Still, I think this is the best possible solution, provided it's well documented. If we find that Guava changes this in the future we can consider other options.

Commit:
Merge branch 'master' into SPARK-4809

Conflicts:
	assembly/pom.xml
	core/pom.xml

Commit:
Shade Guava in the network-common artifact.

This makes all Spark-generated artifacts use the Guava defined in
the build file at runtime too. The leaked classes ("Optional" and
friends) are now exposed in a few places where they weren't before,
but hopefully that won't cause issues.

SparkQA commented:
[Test build #26199 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26199/consoleFull) for   PR 3658 at commit [`3c93e42`](https://github.com/apache/spark/commit/3c93e42a5e9474b33aa53f7fd6f22998d44a8c52).
 * This patch merges cleanly.

pwendell commented:
LGTM pending tests. @vanzin do you have more work you'd like to do on this or are you happy with it?

SparkQA commented:
[Test build #26199 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26199/consoleFull) for   PR 3658 at commit [`3c93e42`](https://github.com/apache/spark/commit/3c93e42a5e9474b33aa53f7fd6f22998d44a8c52).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26199/
Test PASSed.

vanzin commented:
No, I'm done with it. Thanks for taking a look.

caneGuy commented on streaming/pom.xml:
```diff
@@ -105,6 +105,14 @@
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-shade-plugin</artifactId>
```
why add this?
